### PR TITLE
feat: Introduce `rawHeaders`

### LINKED ISSUE #2415: The pseudo-header `:status` is included in h2 fetch.
## Bug Description

The pseudo-header `:status` can be obtained by h2 fetch.
A fix for this problem is needed in #2410.

## Reproducible By

There is a simple test.

```js
// test.js
(async () => {
  const http2 = require("node:http2");

  const undici = require("undici");

  const serverOptions = {
    key:
      "-----BEGIN RSA PRIVATE KEY-----\r\n" +
      "MIICXAIBAAKBgQCqK4GJEFelA5sWipPsBgONPuwxAL3io9GexKlTv6YPsXkfbgkP\r\n" +
      "a8P6zbzqt4eOVjFtDyqrMkJDmc4gRqGNk3+VzOPqaTLcrvOgDAFfY9eF+Z4KmVut\r\n" +
      "JYTYE/3antTOtcP5yAfVVxkpErvvNhE+0+Vtro6FpfGu93iWGsALtM6p5QIDAQAB\r\n" +
      "AoGAfjyB/pTR/8baAPHSZA/eCKT6ACwlsQCH23jWiBsu7ItiaA9vxzHRk7eWAqUd\r\n" +
      "eVi1l8rdg1xoTc/GWqkq5WTbCqbfBE23UQF65wXcBg+sB7aPcpK48PsnIKM3x0ty\r\n" +
      "Xe33hkSQUn9WzH6TTN0b9pFOHb66bHpmY0bM+3g6xxA4ByECQQDTZuw3opoKgt8t\r\n" +
      "tVNnOeOCgcqwIsZZAv97WhKLPJw1iVj5zSJA+fOnskGHevHBByhTPG2Fxjosv4kB\r\n" +
      "wbFAatktAkEAzhHIjyLWG55LF8V3wqlb9IQj4ufy9Oz+GgUSfU83+w2ZsPRW8ywY\r\n" +
      "Nn40dZe4H3cgjPg2vF2Iw3R0QtafWw0WmQJBAJDfR6IY3Fe3uBoq2S00jqUxLSXn\r\n" +
      "ruXxGAQAhhTFRIeryD3oRkEhCckyb5XYuzP9I7VnA+dZISTpXYrGoiHzF0ECQFGG\r\n" +
      "p5o5Zgu3LG5TKXZSGpc6UBhNO8m6mS0iJJZ+6N7iGpPvlxbNsyab95UuWSHZt38O\r\n" +
      "FFmhh6IMSJLI5nLXUsECQBjEDhjqK533gETR+1pF6qIoKxCYWJHe2s7lRUNnEgty\r\n" +
      "CRlOuuMe4yigwWVHjM4914FujY81uqrsmf36H17hWxo=\r\n" +
      "-----END RSA PRIVATE KEY-----\r\n",
    cert:
      "-----BEGIN CERTIFICATE-----\r\n" +
      "MIIClTCCAf6gAwIBAgIJV+VFRrTLKIw6MA0GCSqGSIb3DQEBBQUAMGkxFDASBgNV\r\n" +
      "BAMTC2V4YW1wbGUub3JnMQswCQYDVQQGEwJVUzERMA8GA1UECBMIVmlyZ2luaWEx\r\n" +
      "EzARBgNVBAcTCkJsYWNrc2J1cmcxDTALBgNVBAoTBFRlc3QxDTALBgNVBAsTBFRl\r\n" +
      "c3QwHhcNMjMxMTA5MDgwOTI2WhcNMjQxMTA4MDgwOTI2WjBpMRQwEgYDVQQDEwtl\r\n" +
      "eGFtcGxlLm9yZzELMAkGA1UEBhMCVVMxETAPBgNVBAgTCFZpcmdpbmlhMRMwEQYD\r\n" +
      "VQQHEwpCbGFja3NidXJnMQ0wCwYDVQQKEwRUZXN0MQ0wCwYDVQQLEwRUZXN0MIGf\r\n" +
      "MA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQCqK4GJEFelA5sWipPsBgONPuwxAL3i\r\n" +
      "o9GexKlTv6YPsXkfbgkPa8P6zbzqt4eOVjFtDyqrMkJDmc4gRqGNk3+VzOPqaTLc\r\n" +
      "rvOgDAFfY9eF+Z4KmVutJYTYE/3antTOtcP5yAfVVxkpErvvNhE+0+Vtro6FpfGu\r\n" +
      "93iWGsALtM6p5QIDAQABo0UwQzAMBgNVHRMEBTADAQH/MAsGA1UdDwQEAwIC9DAm\r\n" +
      "BgNVHREEHzAdhhtodHRwOi8vZXhhbXBsZS5vcmcvd2ViaWQjbWUwDQYJKoZIhvcN\r\n" +
      "AQEFBQADgYEAWxD2Q8JRJQC+O7tMSoU5M2Ko5VlCIpgWfqQxrdJ1581izZLtie8Y\r\n" +
      "ccYVYYqoXNnIiE7K/m45UAgN7FSeWGmoCD5GHd8wzOOlmWT0W7+cCufUbfnUsa4g\r\n" +
      "KE7vqFuJjXxtC9iU6sQcYPXk6zVQpCpkUl1Ib5Xsk8Qevk3xvfJR/Vo=\r\n" +
      "-----END CERTIFICATE-----\r\n",
  };

  const server = http2.createSecureServer(serverOptions);

  server.on("stream", (stream, headers) => {
    stream.respond({
      ":status": 200,
    });
    stream.end("test");
  });

  // waiting start server
  await /** @type {Promise<void>} */ (
    new Promise((resolve) => server.listen(3000, "0.0.0.0", () => resolve()))
  );

  const client = new undici.Client(
    `https://localhost:${server.address().port}`,
    {
      connect: {
        rejectUnauthorized: false,
      },
      allowH2: true,
    }
  );

  const response = await undici.fetch(
    `https://localhost:${server.address().port}/`,
    {
      method: "GET",
      dispatcher: client,
    }
  );

  server.close();

  await client.close();

  /**
   * @param {() => any} fn
   * @returns {any}
   */
  function call(fn) {
    try {
      return fn();
    } catch (e) {
      return e;
    }
  }

  /**
   * @param {boolean} cond
   * @param {string | Error} message
   */
  function assert(cond, message) {
    if (!cond) console.error(message);
  }

  assert(
    call(() => response.headers.get(":status")) !== "200",
    "Assertion: It should not be possible to get the pseudo-headers `:status`."
  );

  for (const key of response.headers.keys()) {
    assert(
      !key.startsWith(":"),
      "Assertion: The pseudo-headers `" +
        key +
        "` must not be included in `Headers#keys`."
    );
  }

  console.log([...response.headers]);
})();
```

## Expected Behavior

The above test should succeed.

The pseudo-header `:status` is not passed to `Headers`.

## Logs & Screenshots

```shell
> $ node test.js
(node:7544) [UNDICI-H2] Warning: H2 support is experimental, expect them to change at any time.
(Use `node --trace-warnings ...` to show where the warning was created)
Assertion: It should not be possible to get the pseudo-headers `:status`.
Assertion: The pseudo-headers `:status` must not be included in `Headers#keys`.
[ [ ':status', '200' ], [ 'date', 'Thu, 09 Nov 2023 09:10:56 GMT' ] ]
```

## Environment

Windows 11
Node.js v21.1.0
Undici v5.27.2

### Additional context

related to: #2409 #2410

### PR DESCRIPTION
**Pull request by tsctx**
## This relates to...

Fixes #2415

## Changes

See #2415

## Status

- [ ] I have read and agreed to the [Developer's Certificate of Origin][cert]
- [ ] Tested
- [ ] Benchmarked (**optional**)
- [ ] Documented
- [ ] Review ready
- [ ] In review
- [ ] Merge ready

[cert]: https://github.com/nodejs/undici/blob/main/CONTRIBUTING.md

### CODE REVIEW

**Comment by tsctx** [comment]
Hi @KhafraDev @metcoder95,
All current tests have passed on my machine.
I still have some things to do (add tests), etc., but can you take a look at it?
Do you want to add `rawHeaders` to `onInfo` as well?

**Comment by KhafraDev** [comment]
Parsing the headers readily will majorly slow down undici

**Comment by metcoder95** [comment]
It is quite different from what I would do 🤔 
Have you considered just altering this part? https://github.com/nodejs/undici/blob/ba4ca327843de62a83c1f9c32acc303bd6b8545f/lib/client.js#L1799-L1801

I would like to keep the `h2` specifics as close as possible to the `h2` only code (more specifically `writeh2`.
My idea was to use the `statusText` and extend it to also support an object.
If `h2`, it receives the `rawHeaders` from the response, if `http/1.1`it remains as `string`.

By following the approach I suggest, I won't see much of performance penalties, considering that the `client` is scoped to a single `origin`, so if the `origin` accepts `h2` and is enabled, it will just remains as `h2`, and same for `http/1.1`.

**Comment by tsctx** [comment]
@metcoder95 
Yes, I thought about changing it there too, but quit because I thought it would add too many arguments.
I would like to do it your way.

**Comment by tsctx** [comment]
All that remains is to add tests and modify the documentation.

**Comment by tsctx** [comment]
Passed all current and added tests. Could you please look at it again?
cc @KhafraDev @metcoder95

File: lib/api/api-pipeline.js [hunk 1]
```diff
@@ -157,13 +157,13 @@ class PipelineHandler extends AsyncResource {
     this.context = context
   }
 
-  onHeaders (statusCode, rawHeaders, resume) {
+  onHeaders (statusCode, rawHeaders, resume, statusMessageOrRawHeaders) {
```
Comment (KhafraDev) [inline lib/api/api-pipeline.js:160]: onHeaders can now have 2 rawHeaders.

File: lib/api/api-request.js [hunk 2]
```diff
@@ -77,14 +77,14 @@ class RequestHandler extends AsyncResource {
     this.context = context
   }
 
-  onHeaders (statusCode, rawHeaders, resume, statusMessage) {
+  onHeaders (statusCode, rawHeaders, resume, statusMessageOrRawHeaders) {
```
Comment (KhafraDev) [inline lib/api/api-request.js:80]: same here

**Review by KhafraDev — COMMENTED** [review]
I agree with @metcoder95. This solution seems overly complicated and not correct; there shouldn't be a need to change the non-http2 parts.

File: lib/api/api-pipeline.js (hunk 1 above)
Comment (tsctx) [reply lib/api/api-pipeline.js:160]: The name of the variable is incorrect.

**Pull request closed without merging**